src/service-worker/core.js

```diff
@@ -163,6 +163,38 @@ export function isBlacklisted(version) {
 }
 
 /**
+ * A mapping from a client's referrer into the time that referrer last made a
+ * request. This is used as a fallback to a clientId for Foreign Fetch, since
+ * it does not provide a unique clientId.
+ *
+ * @type {!Object<string, number>}
+ */
+const referrersLastRequestTime = Object.create(null);
```

Comment:
Hopefully this doesn't get too big. But shouldn't in practice. Maybe leave comment to that extend.